test/distributed/org/apache/cassandra/distributed/test/accord/AccordMetricsTest.java

```diff
@@ -0,0 +1,258 @@
+
+    @Override
+    protected Logger logger()
+    {
+        return logger;
+    }
+
+    @BeforeClass
+    public static void setupClass() throws IOException
+    {
+        AccordTestBase.setupClass();
+        SHARED_CLUSTER.forEach(node -> node.runOnInstance(() -> AccordService.instance().createEpochFromConfigUnsafe()));
+        SHARED_CLUSTER.forEach(node -> node.runOnInstance(() -> AccordService.instance().setCacheSize(0)));
+        for (int i = 0; i < SHARED_CLUSTER.size(); i++) // initialize metrics
+            logger.trace(SHARED_CLUSTER.get(i + 1).callOnInstance(() -> AccordMetrics.readMetrics.toString() + AccordMetrics.writeMetrics.toString()));
+    }
+
+    String writeCql()
+    {
+        return "BEGIN TRANSACTION\n" +
+               "  LET val = (SELECT v FROM " + currentTable + " WHERE k=? AND c=?);\n" +
+               "  SELECT val.v;\n" +
+               "  UPDATE " + currentTable + " SET v = v + 1 WHERE k=? AND c=?;\n" +
+               "COMMIT TRANSACTION";
+    }
+
+    String readCql()
+    {
+        return "BEGIN TRANSACTION\n" +
+               "  LET val = (SELECT v FROM " + currentTable + " WHERE k=? AND c=?);\n" +
+               "  SELECT val.v;\n" +
+               "COMMIT TRANSACTION";
+    }
+
+    Map<Integer, Map<String, Long>> countingMetrics0;
+
+    @Before
+    public void beforeTest()
+    {
+        SHARED_CLUSTER.filters().reset();
+        SHARED_CLUSTER.schemaChange("CREATE TABLE " + currentTable + " (k int, c int, v int, PRIMARY KEY (k, c))");
+        SHARED_CLUSTER.coordinator(1).execute("INSERT INTO " + currentTable + " (k, c, v) VALUES (0, 0, 0)", ConsistencyLevel.ALL);
+    }
+
+    @Test
+    public void testRegularMetrics() throws Exception
+    {
+        countingMetrics0 = getMetrics();
+        SHARED_CLUSTER.coordinator(1).executeWithResult(writeCql(), ConsistencyLevel.ALL, 0, 0, 0, 0);
+        assertCoordinatorMetrics(0, "rw", 1, 0, 0, 0, 0);
+        assertCoordinatorMetrics(1, "rw", 0, 0, 0, 0, 0);
+        assertReplicaMetrics(0, "rw", 1, 1, 1);
+        assertReplicaMetrics(1, "rw", 1, 1, 1);
+        assertZeroMetrics("ro");
+
+        countingMetrics0 = getMetrics();
+        SHARED_CLUSTER.coordinator(1).executeWithResult(readCql(), ConsistencyLevel.ALL, 0, 0);
+        assertCoordinatorMetrics(0, "ro", 1, 0, 0, 0, 0);
+        assertCoordinatorMetrics(1, "ro", 0, 0, 0, 0, 0);
+        assertReplicaMetrics(0, "ro", 1, 1, 0);
+        assertReplicaMetrics(1, "ro", 1, 1, 0);
+        assertZeroMetrics("rw");
+    }
+
+    @Test
+    public void testPreemptionMetrics()
+    {
+        IMessageFilters.Filter commitFilter = SHARED_CLUSTER.filters().outbound().verbs(Verb.ACCORD_COMMIT_REQ.id).from(1).to(2).drop();
+        commitFilter.on();
+
+        countingMetrics0 = getMetrics();
+        try
+        {
+            SHARED_CLUSTER.coordinator(1).executeWithResult(writeCql(), ConsistencyLevel.ALL, 0, 0, 0, 0);
+        }
+        catch (RuntimeException t)
```

Comment:
nit: I wonder if it would make more sense to catch `CassandraException` in places like this, so things like NPE just fail without even worrying about the following assertion...